### PR TITLE
fix sgen if/else issue

### LINKED ISSUE #25248: Fix the SGEN performance issue
The app that use SGEN take much longer than the app not use SGEN. This is caused by the super long “if…else if …else if” statement in the Wrte_Object/Read_Object methods.  Need port the fix in SG to corefx.

### PR DESCRIPTION
**Pull request by huanwu**
Port the if/else fix from SG to corefx. Fix another if/else issue when write member elements only when the total number of members is bigger than 1000.

Fix #25248 

@shmao @zhenlan @mconnew

### CODE REVIEW

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs [hunk 1]
```diff
@@ -2197,6 +2200,8 @@ private void WriteMemberElementsIf(Member[] members, Member anyElement, string e
                 Writer.WriteLine("switch (state) {");
             }
             int cases = 0;
+            bool largeNumberOfMembers = (members.Length > 1000);
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs]: Why is the number set to 1000? 

Besides, I think it'd be better to exact the value into a readonly variable.

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs [hunk 2]
```diff
@@ -2197,6 +2200,8 @@ private void WriteMemberElementsIf(Member[] members, Member anyElement, string e
...
+            bool useDoWhile = largeNumberOfMembers && !isSequence;
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs]: We might want to add comments here explaining why we want to use do...while.

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs (hunk 1 above)
Comment (zhenlan) [reply src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs]: Unless there are perf concerns, I would not bother to generate different code.

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs [hunk 3]
```diff
@@ -2217,7 +2220,8 @@ private void WriteMemberElementsIf(Member[] members, Member anyElement, string e
                     ElementAccessor e = elements[j];
                     string ns = e.Form == XmlSchemaForm.Qualified ? e.Namespace : "";
                     if (!isSequence && e.Any && (e.Name == null || e.Name.Length == 0)) continue;
-                    if (!firstElement || (!isSequence && count > 0))
+
+                    if(!firstElement && isSequence)
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs]: I don't think there's any case we want to write `else`. And when would this condition be `true`?

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs [hunk 4]
```diff
@@ -2344,7 +2353,10 @@ private void WriteMemberElementsIf(Member[] members, Member anyElement, string e
                 if (isSequence)
                     Writer.WriteLine("default:");
                 else
-                    Writer.WriteLine("else {");
+                {
+                    Writer.WriteLine("Finish:");
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs]: Is this the right place for `Finish:`? This seems to be the line of original `else`. I guessed the `Finish` label should be after the original `else` block.

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs [hunk 5]
```diff
@@ -2353,9 +2365,13 @@ private void WriteMemberElementsIf(Member[] members, Member anyElement, string e
                 if (isSequence)
                 {
                     Writer.WriteLine("break;");
+                    Writer.Indent--;
+                    Writer.WriteLine("}");
+                }
+                else
+                {
+                    Writer.Indent--;
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs]: The `Finish` label should probably be here.

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs (hunk 3 above)
Comment (huanwu) [reply src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs]: All the changes are only for the case when isSequence = false.

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs (hunk 4 above)
Comment (huanwu) [reply src/Microsoft.XmlSerializer.Generator/src/Microsoft/XmlSerializer/Generator/XmlSerializationReader.cs]: You are right. So the Finish tag has to put outside of this method. Seems goto is not the right way in this case and I will switch back to do while(false).

**Comment by stephentoub** [comment]
@huanwu, @shmao, what's the status of this PR?  Is there still work to be done, or can it be squashed and merged?

**Pull request merged**